BRWallet.c

```diff
@@ -635,35 +740,53 @@ BRTransaction *BRWalletCreateTxForOutputs(BRWallet *wallet, const BRTxOutput out
-    for (i = 0; tx && i < tx->inCount; i++) {
-        for (j = (uint32_t)array_count(wallet->internalChain); j > 0; j--) {
-            if (BRAddressEq(tx->inputs[i].address, &wallet->internalChain[j - 1])) internalIdx[internalCount++] = j - 1;
+    // First we check the witness address chain
+    for (i = 0; i < tx->inCount; i++) {
+        for (j = (uint32_t)array_count(wallet->changeWitnessAddrs); j > 0; j--) {
+            if (BRAddressEq(tx->inputs[i].address, &wallet->changeWitnessAddrs[j - 1])) changeIdx[changeCount++] = j - 1;
         }
 
-        for (j = (uint32_t)array_count(wallet->externalChain); j > 0; j--) {
-            if (BRAddressEq(tx->inputs[i].address, &wallet->externalChain[j - 1])) externalIdx[externalCount++] = j - 1;
+        for (j = (uint32_t)array_count(wallet->receiveWitnessAddrs); j > 0; j--) {
+            if (BRAddressEq(tx->inputs[i].address, &wallet->receiveWitnessAddrs[j - 1])) receiveIdx[receiveCount++] = j - 1;
         }
+
     }
 
-    pthread_mutex_unlock(&wallet->lock);
+    // Then we check the legacy address chain
+    for (i = 0; i < tx->inCount; i++) {
+        for (j = (uint32_t)array_count(wallet->changeAddrs); j > 0; j--) {
+            if (BRAddressEq(tx->inputs[i].address, &wallet->changeAddrs[j - 1])) changeLegacyIdx[changeLegacyCount++] = j - 1;
+        }
 
-    BRKey keys[internalCount + externalCount];
+        for (j = (uint32_t)array_count(wallet->receiveAddrs); j > 0; j--) {
+            if (BRAddressEq(tx->inputs[i].address, &wallet->receiveAddrs[j - 1])) receiveLegacyIdx[receiveLegacyCount++] = j - 1;
+        }
+    }
+
+    pthread_mutex_unlock(&wallet->lock);
 
     if (seed) {
-        BRBIP32PrivKeyList(keys, internalCount, seed, seedLen, SEQUENCE_INTERNAL_CHAIN, internalIdx);
-        BRBIP32PrivKeyList(&keys[internalCount], externalCount, seed, seedLen, SEQUENCE_EXTERNAL_CHAIN, externalIdx);
+        BRKey keys[changeCount + receiveCount + changeLegacyCount + receiveLegacyCount];
+
+        BRBIP32PrivKeyList(keys, changeCount, seed, seedLen, SEQUENCE_CHANGE_WITNESS_CHAIN, changeIdx);
+        BRBIP32PrivKeyList(&keys[changeCount], receiveCount, seed, seedLen, SEQUENCE_RECEIVE_WITNESS_CHAIN, receiveIdx);
+        BRBIP32PrivKeyList(&keys[changeCount + receiveCount], changeLegacyCount, seed, seedLen, SEQUENCE_CHANGE_CHAIN, changeLegacyIdx);
+        BRBIP32PrivKeyList(&keys[changeCount + receiveCount + changeLegacyCount], receiveLegacyCount, seed, seedLen, SEQUENCE_RECEIVE_CHAIN, receiveLegacyIdx);
+
+        if (tx) r = BRTransactionSign(tx, forkId, keys, changeCount + receiveCount + changeLegacyCount + receiveLegacyCount);
+        for (i = 0; i < changeCount + receiveCount + changeLegacyCount + receiveLegacyCount; i++) BRKeyClean(&keys[i]);
+
         // TODO: XXX wipe seed callback
         seed = NULL;
-        if (tx) r = BRTransactionSign(tx, forkId, keys, internalCount + externalCount);
-        for (i = 0; i < internalCount + externalCount; i++) BRKeyClean(&keys[i]);
+
```

Comment:
(Completed review down to here as of 5457223000549d129ccf213160690d2741211cdb; submitting this so you can see my comments sooner rather than later, and will pick up again from here later; did not review tests)

Reply:
@luke-jr Finished reading, the tests, I figured out later that they weren't really checking if the transactions was signed or not.

Please only view the current tip of the segwit branch, I remove the transaction test and only left some tests for addresses recently.

Reply:
Despite the GitHub timestamps, I actually did most of the review I posted last week when it was the current tip. I'll be using the new tip for review going forward.

Reply:
That makes sense. Thank you!